fairlearn/grid_search.py

```diff
@@ -0,0 +1,144 @@
+            if current_a == 0:
+                w = 2*current_y - 1 - L * p_ratio
+            else:
+                w = 2*current_y - 1 + L
+
+            weights.append(w)
+
+        return np.array(weights)
+
+    def grid_search_binary_protected_attribute(self, learner, x, y, protected_attribute, lagrangian_multipliers=None, number_lagrangian_multipliers=11):
+        """Function to generate a list of models for a classification problem with
+        a single binary protected attribute. The models sweep through different potential
+        Lagrangian multipliers for the constrained optimisation problem (the constraint
+        being demographic parity), each one corresponding to a particular tradeoff between
+        fairness and accuracy
+
+        :param learner: An object which can be used to fit a model to features, labels and
+        weights. A deep copy of this is made for each value of the Lagrangian multiplier used
+        :type learner: Must implement a fit(x, y, sample_weight) method
+
+        :param x: The array of training data features (which may or may not contain the
+        protected attribute). Must have as many rows as y and protected_attribute
+        :type x: Numpy array with two dimensions or pandas dataframe
+
+        :param y: The list of binary classes, which must be 0 or 1. Must contain the same
+        number of entries as rows in x
+        :type y: Numpy array with one dimension
+
+        :param protected_attribute: The protected attribute corresponding to each row of x.
+        Must be either 0 or 1
+        :type protected_attribute: Numpy array with one dimension
+        
+        :param lagrangian_multipliers: User specified set of Lagrangian multipliers to use for
+        the optimisation problem. If this is set then num_Ls must be None. The result array will
+        be equal in length to this array
+        :type lagrangian_multipliers: List of real numbers
+        
+        :param number_lagrangian_multipliers: Specifies the number of Lagrangian multipliers to
+        use in the optimisation problem. If this is set then lagrangian_multipliers must be None.
+        The result array will have as many entries as specified here
+
+        :return: The models corresponding to each value of the Lagrangian multiplier tested
+        :rtype: List of dictionaries. Each dictionary has fields "lagrangian_multiplier" and "model." Each
+        model will correspond to the input parameter learner after calling 'fit' on it (a deep
+        copy is made). The user is responsible for converting these objects to an actual model,
+        if further processing is required.
+        """
+        # Must specify either Ls or num_Ls
+        if not (lagrangian_multipliers is None) ^ (number_lagrangian_multipliers is None):
+            raise RuntimeError("Must specify either lagrangian_multipliers or number_lagrangian_multipliers")
+
+        # Check that x, y and protected_attribute have the same number of rows
```

Comment:
I think we should be able to handle a and y represented as either series / 1d arrays, or also dataframes / 2d arrays (where the second dimension=1).

Reply:
I've just added tests which ensure that we use regular lists, numpy or pandas for X, Y and A.  Allowing extra dimensions of length 1 looks to be rather hairer. For example, if we allow Y or A to be ndarrays of size (n,1), should we also allow them to have size (1,n). How about (1, 1, n, 1, 1, 1)?

Reply:
For now, I'm not going to make the dimension checks; the learner itself will scream if it can't coerce the arrays into the right shape. It is something to be addressed in due course, but this PR has already been through a lot